File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/message/digitaltwin/ReadEventsRequest.java
package org.opcfoundation.webserver.service.message.digitaltwin;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.structured.EventFilter;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadEventDetails;
import org.eclipse.milo.opcua.stack.core.types.structured.SimpleAttributeOperand;
import org.opcfoundation.webserver.types.digitaltwin.EventQuery;
import org.opcfoundation.webserver.types.digitaltwin.ObjectServiceContext;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;

public class ReadEventsRequest {
    private final String id;
    private final LocalDateTime startTime;
    private final LocalDateTime endTime;

    private final List<String> select;
    private final EventQuery where;
    private final boolean ascendingOrder;
    private final int limit;
    private final int offset;

    private final ObjectServiceContext context;

    public ReadEventsRequest(
            ObjectServiceContext context,
            LocalDateTime startTime,
            LocalDateTime endTime,
            List<String> select,
            EventQuery where,
            boolean ascendingOrder,
            int limit,
            int offset)
    {
        this.id = context.getObjectId().getId();
        this.context = context;
        this.startTime = startTime;
        this.endTime = endTime;
        this.select = select;
        this.where = where;
        this.ascendingOrder = ascendingOrder;
        this.limit = limit;
        this.offset = offset;
    }

    public String getId() {
        return id;
    }

    public LocalDateTime getStartTime() {
        return startTime;
    }

    public LocalDateTime getEndTime() {
        return endTime;
    }

    public List<String> getSelect() {
        return select;
    }

    public EventQuery getWhere() {
        return where;
    }

    public Boolean getAscendingOrder() {
        return ascendingOrder;
    }

    public int getLimit() {
        return limit;
    }

    public int getOffset() {
        return offset;
    }

    public ObjectServiceContext getContext() {
        return context;
    }

    public static ReadEventsRequest getRequest(ObjectServiceContext context, ReadEventDetails details, int offset) throws UaRuntimeException
    {
        LocalDateTime startTime = details.getStartTime().getJavaInstant().atZone(ZoneOffset.UTC).toLocalDateTime();
        LocalDateTime endTime = details.getEndTime().getJavaInstant().atZone(ZoneOffset.UTC).toLocalDateTime();
        boolean ascendingOrder = true;

        List<String> select = new ArrayList<>();
        EventQuery where;

        if (endTime.isBefore(startTime))
        {
            ascendingOrder = false;
            LocalDateTime tmp = endTime;
            endTime = startTime;
            startTime = tmp;
        }

        EventFilter eventFilter = details.getFilter();

        // Select
        if (null == eventFilter.getSelectClauses() || 0 == eventFilter.getSelectClauses().length) throw new UaRuntimeException(StatusCodes.Bad_ContentFilterInvalid);

        for (SimpleAttributeOperand item: eventFilter.getSelectClauses())
        {
            if (null == item.getBrowsePath() || 0 == item.getBrowsePath().length || 2 < item.getBrowsePath().length) throw new UaRuntimeException(StatusCodes.Bad_ContentFilterInvalid);

            String fieldName = item.getBrowsePath()[0].getName();
            if (null == fieldName || fieldName.isEmpty()) throw new UaRuntimeException(StatusCodes.Bad_ContentFilterInvalid);

            if (2 == item.getBrowsePath().length)
            {
                String subFieldName = item.getBrowsePath()[1].getName();
                if (null == subFieldName || subFieldName.isEmpty()) throw new UaRuntimeException(StatusCodes.Bad_ContentFilterInvalid);
                fieldName += "/";
                fieldName += subFieldName;
            }

            select.add(fieldName);
        }

        if (select.size() != eventFilter.getSelectClauses().length) throw new UaRuntimeException(StatusCodes.Bad_ContentFilterInvalid);

        // Where
        where = EventQuery.getQuery(eventFilter.getWhereClause());

        return new ReadEventsRequest(
                context,
                startTime,
                endTime,
                select,
                where,
                ascendingOrder,
                details.getNumValuesPerNode().intValue(),
                offset);
    }
}

File: opcua-webservice-node/source/src/server/service/message/digitialtwin/ReadEventsRequest.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaError, UaQuery, UaReadEventDetails } from "opcua-webapi-ts";
import { ObjectServiceContext } from "../../../types/digitaltwin/ObjectServiceContext";

export class ReadEventsRequest {
    private readonly _id: string;
    private readonly _startTime: Date;
    private readonly _endTime: Date;
    private readonly _select: Array<string>;
    private readonly _where: UaQuery | null;
    private readonly _ascendingOrder: boolean;
    private readonly _limit: number;
    private readonly _offset: number;
    private readonly _context: ObjectServiceContext;

    constructor(
        context: ObjectServiceContext,
        startTime: Date,
        endTime: Date,
        select: Array<string>,
        where: UaQuery | null,
        ascendingOrder: boolean,
        limit: number,
        offset: number,
    ) {
        this._id = context.objectId.id;
        this._context = context;
        this._startTime = startTime;
        this._endTime = endTime;
        this._select = select;
        this._where = where;
        this._ascendingOrder = ascendingOrder;
        this._limit = limit;
        this._offset = offset;
    }

    get id(): string {
        return this._id;
    }

    get startTime(): Date {
        return this._startTime;
    }

    get endTime(): Date {
        return this._endTime;
    }

    get select(): Array<string> {
        return this._select;
    }

    get where(): UaQuery | null {
        return this._where;
    }

    get ascendingOrder(): boolean {
        return this._ascendingOrder;
    }

    get limit(): number {
        return this._limit;
    }

    get offset(): number {
        return this._offset;
    }

    get context(): ObjectServiceContext {
        return this._context;
    }

    static getRequest(
        context: ObjectServiceContext,
        details: UaReadEventDetails,
        offset: number,
    ): ReadEventsRequest {
        let startTime = details.startTime;
        let endTime = details.endTime;
        let ascendingOrder = true;

        if (endTime.getTime() < startTime.getTime()) {
            ascendingOrder = false;
            const tmp = endTime;
            endTime = startTime;
            startTime = tmp;
        }

        const eventFilter = details.filter;
        const selectClauses = eventFilter.selectClauses;
        if (selectClauses.length === 0) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadContentFilterInvalid));
        }

        let selects : Array<string> = [];
        for (const item of selectClauses) {
            const fieldName = ReadEventsRequest.getSelectField(item.browsePath);
            selects.push(fieldName);
        }
        
        const where = eventFilter.whereClause ? UaQuery.fromContentFilter(eventFilter.whereClause) : null;

        return new ReadEventsRequest(
            context,
            startTime,
            endTime,
            selects,
            where,
            ascendingOrder,
            details.numValuesPerNode,
            offset,
        );
    }

    private static getSelectField(browsePath: Array<string>): string {
        if (browsePath.length === 0 || browsePath.length > 2) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadContentFilterInvalid));
        }

        let fieldName = browsePath[0];
        if (browsePath.length === 2) {
            let subFieldName = browsePath[1];
            if (null == subFieldName || subFieldName.length == 0) {
                throw new UaError(makeUaStatusCode(StatusCodes.BadContentFilterInvalid));
            }
            fieldName = `${fieldName}/${subFieldName}`;
        }  

        return fieldName;
    }
}